### PR TITLE
oonimkall: add wrappers for /test-list/urls

### PR DESCRIPTION
**Pull request by lorenzoPrimi**
Implementing https://github.com/ooni/probe/issues/1331

### CODE REVIEW

File: pkg/oonimkall/session.go [hunk 1]
```diff
@@ -375,3 +377,81 @@ func (sess *Session) CheckIn(ctx *Context, config *CheckInConfig) (*CheckInInfo,
...
+
+// Size returns the number of URLs.
+func (ckw *URLListResult) Size() int64 {
+	return int64(len(ckw.Results))
+}
+
+// URLListResult contains the URLs returnet from the FetchURL API
+type URLListResult struct {
+	Results []model.URLInfo
+}
```
Comment (bassosimone) [inline pkg/oonimkall/session.go:414]: Please, move the struct definition before its methods

File: pkg/oonimkall/session.go [hunk 2]
```diff
@@ -375,3 +377,81 @@ func (sess *Session) CheckIn(ctx *Context, config *CheckInConfig) (*CheckInInfo,
...
+
+// FetchURLList function is called to fetch the urls list to test
```
Comment (bassosimone) [inline pkg/oonimkall/session.go:416]:
```suggestion
// FetchURLList fetches the list of URLs to test
```

File: pkg/oonimkall/session.go [hunk 3]
```diff
@@ -375,3 +377,81 @@ func (sess *Session) CheckIn(ctx *Context, config *CheckInConfig) (*CheckInInfo,
...
+// In the config argument a probe can specify the Categories to test, limit urls and send the Probe's CC
```
Comment (bassosimone) [inline pkg/oonimkall/session.go:417]: I don't think there is need to describe the config argument here.

File: pkg/oonimkall/session.go [hunk 4]
```diff
@@ -375,3 +377,81 @@ func (sess *Session) CheckIn(ctx *Context, config *CheckInConfig) (*CheckInInfo,
...
+		query.Set("country_code", cfg.CountryCode)
+	}
+	if cfg.Limit > 0 {
+		query.Set("limit", fmt.Sprintf("%d", cfg.Limit))
+	}
+	if len(cfg.Categories) > 0 {
+		query.Set("category_codes", strings.Join(cfg.Categories, ","))
+	}
+	var response URLListResult
+	err = psc.GetJSONWithQuery(ctx.ctx, "/api/v1/test-list/urls", query, &response)
```
Comment (bassosimone) [inline pkg/oonimkall/session.go:450]: This code does not strike me as wrong, but there should be already code in `probeservices` for doing that. I'm ~sure you are duplicating existing code here. Can you perhaps could check?

File: pkg/oonimkall/session_integration_test.go [hunk 5]
```diff
@@ -419,6 +419,67 @@ func TestCheckInNoParams(t *testing.T) {
 	}
 }
 
+func TestFetchURLListSucc(t *testing.T) {
```
Comment (bassosimone) [inline pkg/oonimkall/session_integration_test.go:422]:
```suggestion
func TestFetchURLListSuccess(t *testing.T) {
```

File: pkg/oonimkall/session_integration_test.go [hunk 6]
```diff
@@ -419,6 +419,67 @@ func TestCheckInNoParams(t *testing.T) {
...
+		t.Fatal(err)
+	}
+	ctx := sess.NewContext()
+	config := oonimkall.URLListConfig{
+		Limit: 10,
+	}
+	config.AddCategory("NEWS")
+	config.AddCategory("CULTR")
+	result, err := sess.FetchURLList(ctx, &config)
+
```
Comment (bassosimone) [inline pkg/oonimkall/session_integration_test.go:434]:
```suggestion
```

File: pkg/oonimkall/session_integration_test.go [hunk 7]
```diff
@@ -419,6 +419,67 @@ func TestCheckInNoParams(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("unexpected error: %+v", err)
+	}
+	if result == nil || result.Results == nil {
+		t.Fatal("got nil result")
+	}
+	if len(result.Results) < 1 {
```
Comment (bassosimone) [inline pkg/oonimkall/session_integration_test.go:441]: `len(result.Results)` is zero when `result.Results` is `nil`, so you can write more compact code above.

File: pkg/oonimkall/session_integration_test.go [hunk 8]
```diff
@@ -419,6 +419,67 @@ func TestCheckInNoParams(t *testing.T) {
...
+	config.AddCategory("NEWS")
+	config.AddCategory("CULTR")
+	result, err := sess.FetchURLList(ctx, &config)
+	if err != nil {
+		t.Fatalf("unexpected error: %+v", err)
+	}
+	if result == nil || result.Results == nil {
+		t.Fatal("got nil result")
+	}
+	if len(result.Results) < 1 {
```
Comment (bassosimone) [inline pkg/oonimkall/session_integration_test.go:470]: Same comment as before

**Review by bassosimone — CHANGES_REQUESTED** [review]
Thanks! I feel like we should be reusing more code in `probeservices` here!

File: pkg/oonimkall/session.go (hunk 4 above)
Comment (lorenzoPrimi) [reply pkg/oonimkall/session.go:449]: Yes there is similar code in probeservices/urls.go, but I thought we needed to duplicate the code we need for the wrappers

File: pkg/oonimkall/session_integration_test.go (hunk 7 above)
Comment (lorenzoPrimi) [reply pkg/oonimkall/session_integration_test.go:441]: Is there a way to differentiate between nil and empty array? I don't know how you want the code to behave

Comment (bassosimone) [reply pkg/oonimkall/session_integration_test.go:441]: In this case an empty array or a nil array is wrong, so I don't think we need to distinguish.

**Review by bassosimone — APPROVED** [review]
🐳 

The code looks good to me! I am going to test locally and merge later today!

Thank you!

File: pkg/oonimkall/session.go [hunk 9]
```diff
@@ -375,3 +375,69 @@ func (sess *Session) CheckIn(ctx *Context, config *CheckInConfig) (*CheckInInfo,
...
 }
+
+// URLListConfig contains configuration for fetching the URL list.
+type URLListConfig struct {
+	Categories  []string // Categories to query for (empty means all)
+	CountryCode string   // CountryCode is the optional country code
+	Limit       int64    // Max number of URLs (<= 0 means no limit)
+}
+
+// URLListResult contains the URLs returnet from the FetchURL API
```
Comment (bassosimone) [inline pkg/oonimkall/session.go:386]:
```suggestion
// URLListResult contains the URLs returned from the FetchURL API
```

File: pkg/oonimkall/session.go [hunk 10]
```diff
@@ -375,3 +375,69 @@ func (sess *Session) CheckIn(ctx *Context, config *CheckInConfig) (*CheckInInfo,
...
+type URLListResult struct {
+	Results []model.URLInfo
+}
+
+// AddCategory adds category code to the array in URLListConfig
+func (ckw *URLListConfig) AddCategory(cat string) {
+	ckw.Categories = append(ckw.Categories, cat)
+}
+
+// At gets the URLInfo at position idx from CheckInInfoWebConnectivity.URLs
```
Comment (bassosimone) [inline pkg/oonimkall/session.go:396]:
```suggestion
// At gets the URLInfo at position idx from CheckInInfoWebConnectivity.URLs. It returns
// nil if you are using an outs of bound index.
```

File: pkg/oonimkall/session_integration_test.go [hunk 11]
```diff
@@ -419,6 +419,59 @@ func TestCheckInNoParams(t *testing.T) {
...
+	found := false
+	for _, entry := range result.Results {
+		if entry.CountryCode == "IT" {
+			found = true
+		}
+	}
+	if !found {
+		t.Fatalf("not found url for country code: IT")
+	}
+}
```
Comment (bassosimone) [inline pkg/oonimkall/session_integration_test.go:474]:
```suggestion
}

```

**Comment by bassosimone** [comment]
I am going to merge this PR as soon as the automatic tests succeed.

**Pull request merged**